enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj

```diff
@@ -14,6 +14,23 @@
     [toucan.db :as db]
     [toucan.hydrate :refer [hydrate]]))
 
+(defn collection-set-for-user
+  "Given an optional user ID, find the transitive set of all Collection IDs which are either:
+  (a) global (ie. no one's personal collection);
+  (b) owned by this user (when user ID is non-nil); or
+  (c) descended from one of the above."
+  [user-or-nil]
+  (let [roots    (db/select ['Collection :id :location :personal_owner_id] :location "/")
+        unowned  (remove :personal_owner_id roots)
+        owned    (when user-or-nil
+                   (filter #(= user-or-nil (:personal_owner_id %)) roots))
+        top-ids  (->> (concat owned unowned)
+                      (map :id)
+                      set)]
+    (->> (concat unowned owned)
+         (map collection/descendant-ids)
+         (reduce set/union top-ids))))
```

Comment:
It's probably not a huge deal here since this is SerDes and not something that is done all the time, e.g. the QP or an API endpoint, but this is an N+1 query situation. `descendant-ids` is defined like this:

```clj
(s/defn descendant-ids :- (s/maybe #{su/IntGreaterThanZero})
  "Return a set of IDs of all descendant Collections of a `collection`."
  [collection :- CollectionWithLocationAndIDOrRoot]
  (db/select-ids Collection :location [:like (str (children-location collection) \%)]))
```

which means if you fetched 20 collections in the first query then you'd have to do 20 additional database queries to fetch all the of descendant IDs.

I have a better idea: why not just rework `descendant-ids` to accept one **or more** collections? If you pass multiple collections instead of generating SQL like 

```sql
SELECT id
FROM collection
WHERE location LIKE '/1/%';
```

it could do something like

```sql
SELECT id
FROM collection
WHERE location LIKE '/1/%'
  OR location LIKE '/2/%';
```

then you have a total of two application DB queries. And this sort of change is more general, because the new and improved version of `descendant-ids` might be useful elsewhere as well.